File: packages/frontend/src/api/product.api.ts
import api from './client';
import type { ApiResponse, Product, PaginatedResponse, FilterOptions, Category, FeedResponse, TraderGalleryProduct } from '@/types';

export interface ProductQuery {
  search?: string;
  vendorId?: string;
  brandId?: string;
  categoryId?: string;
  pattern?: string;
  fabric?: string;
  color?: string;
  minPrice?: number;
  maxPrice?: number;
  page?: number;
  limit?: number;
}

export const productApi = {
  list: (params: ProductQuery = {}) =>
    api.get<ApiResponse<PaginatedResponse<Product>>>('/products', { params }).then((r) => r.data.data),

  getById: (id: string) =>
    api.get<ApiResponse<Product>>(`/products/${id}`).then((r) => r.data.data),

  /** Trader photo grid only — smaller/faster than `getById`. */
  getTraderGallery: (id: string) =>
    api.get<ApiResponse<TraderGalleryProduct>>(`/products/${id}/gallery`).then((r) => r.data.data),
  getMyProducts: () =>
    api.get<ApiResponse<Product[]>>('/products/my').then((r) => r.data.data),

  create: (data: FormData | Record<string, unknown>) =>
    api.post<ApiResponse<Product>>('/products', data).then((r) => r.data.data),

  update: (id: string, data: Record<string, unknown>) =>
    api.put<ApiResponse<Product>>(`/products/${id}`, data).then((r) => r.data.data),

  delete: (id: string) =>
    api.delete<ApiResponse<null>>(`/products/${id}`).then((r) => r.data),

  bulkDelete: (ids: string[]) =>
    api.post<ApiResponse<{ deleted: number }>>('/products/bulk-delete', { ids }).then((r) => r.data.data),

  bulkUpdate: (ids: string[], updates: { categoryId?: string; moq?: number; status?: string }) =>
    api.put<ApiResponse<{ updated: number }>>('/products/bulk-update', { ids, ...updates }).then((r) => r.data.data),

  getCategories: () =>
    api.get<ApiResponse<Category[]>>('/products/categories').then((r) => r.data.data),

  getFilters: () =>
    api.get<ApiResponse<FilterOptions>>('/products/filters').then((r) => r.data.data),

  feed: (params: { cursor?: string; limit?: number; categoryId?: string } = {}) =>
    api.get<ApiResponse<FeedResponse>>('/products/feed', { params }).then((r) => r.data.data),

  saveProduct: (productId: string) =>
    api.post<ApiResponse<null>>('/products/save', { productId }).then((r) => r.data),

  unsaveProduct: (productId: string) =>
    api.delete<ApiResponse<null>>(`/products/save/${productId}`).then((r) => r.data),

  getSavedProducts: () =>
    api.get<ApiResponse<Product[]>>('/products/saved').then((r) => r.data.data),
};

File: packages/mobile/lib/core/api/product_api.dart
import 'package:dio/dio.dart';

import '../network/api_response.dart';

class ProductApi {
  ProductApi(this._dio);
  final Dio _dio;

  Future<Map<String, dynamic>> list({Map<String, dynamic>? query}) async {
    final res = await _dio.get<Map<String, dynamic>>('/products', queryParameters: query);
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<Map<String, dynamic>> getById(String id) async {
    final res = await _dio.get<Map<String, dynamic>>('/products/$id');
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<Map<String, dynamic>> getTraderGallery(String id) async {
    final res = await _dio.get<Map<String, dynamic>>('/products/$id/gallery');
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<List<dynamic>> getMyProducts() async {
    final res = await _dio.get<Map<String, dynamic>>('/products/my');
    return ApiResponse.unwrapData(res.data!, (d) => d as List<dynamic>);
  }

  Future<Map<String, dynamic>> create(Object data) async {
    final res = await _dio.post<Map<String, dynamic>>('/products', data: data);
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<Map<String, dynamic>> update(String id, Map<String, dynamic> data) async {
    final res = await _dio.put<Map<String, dynamic>>('/products/$id', data: data);
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<void> delete(String id) async {
    await _dio.delete<Map<String, dynamic>>('/products/$id');
  }

  Future<Map<String, dynamic>> bulkDelete(List<String> ids) async {
    final res = await _dio.post<Map<String, dynamic>>('/products/bulk-delete', data: {'ids': ids});
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<Map<String, dynamic>> bulkUpdate(
    List<String> ids,
    Map<String, dynamic> updates,
  ) async {
    final res = await _dio.put<Map<String, dynamic>>(
      '/products/bulk-update',
      data: {...updates, 'ids': ids},
    );
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<List<dynamic>> getCategories() async {
    final res = await _dio.get<Map<String, dynamic>>('/products/categories');
    return ApiResponse.unwrapData(res.data!, (d) => d as List<dynamic>);
  }

  Future<Map<String, dynamic>> getFilters() async {
    final res = await _dio.get<Map<String, dynamic>>('/products/filters');
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<Map<String, dynamic>> feed({String? cursor, int? limit, String? categoryId}) async {
    final res = await _dio.get<Map<String, dynamic>>(
      '/products/feed',
      queryParameters: {
        'cursor': ?cursor,
        'limit': ?limit,
        'categoryId': ?categoryId,
      },
    );
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<void> saveProduct(String productId) async {
    await _dio.post<Map<String, dynamic>>('/products/save', data: {'productId': productId});
  }

  Future<void> unsaveProduct(String productId) async {
    await _dio.delete<Map<String, dynamic>>('/products/save/$productId');
  }

  Future<List<dynamic>> getSavedProducts() async {
    final res = await _dio.get<Map<String, dynamic>>('/products/saved');
    return ApiResponse.unwrapData(res.data!, (d) => d as List<dynamic>);
  }
}
